### ISSUE TITLE
Bug Report: Docker Hub Automated Build of Dockefile Erroring

### ISSUE THREAD

**Issue by matthewfeickert** [diana-hep/pyhf#239]
The pyhf Docker image that [Docker Hub automatically builds](https://hub.docker.com/r/pyhf/pyhf/) on pushes to `master` is erroring.

# Expected Behavior

Docker Hub successfully builds and publishes a Docker image for pyhf

# Actual Behavior

c.f. [build bgwk4gm994woza9reaje5se log](https://hub.docker.com/r/pyhf/pyhf/builds/bgwk4gm994woza9reaje5se/)

The main issue is the following:

```
Directory '.' is not installable. File 'setup.py' not found.
```

# Checklist

- [x] Run `git fetch` to get the most up to date version of `master`
- [x] Searched through existing Issues to confirm this is not a duplicate issue
- [x] Filled out the Description, Expected Behavior, Actual Behavior, and Steps to Reproduce sections above or have edited/removed them in a way that fully describes the issue

**Comment by lukasheinrich** [comment]
since we have the Dockerfile in a subrepo the builds need to happen (and do happen) on Docker Cloud

https://cloud.docker.com/swarm/pyhf/repository/docker/pyhf/pyhf/builds

what's your docker hub username?

**Comment by matthewfeickert** [comment]
@lukasheinrich It is just my GitHub handle: [matthewfeickert](https://hub.docker.com/u/matthewfeickert/)

**Comment by lukasheinrich** [comment]
@matthewfeickert you should have access to the above link now

**Comment by matthewfeickert** [comment]
>since we have the Dockerfile in a subdirectory the builds need to happen (and do happen) on Docker Cloud

Maybe I'm missing the obvious, but couldn't the [path to the Dockerfile for the auto build just be configured in the Docker Hub build settings](https://github.com/diana-hep/pyhf/pull/225#issuecomment-418191924)?

**Comment by matthewfeickert** [comment]
> you should have access to the above link now

Yup! Thanks.

**Comment by lukasheinrich** [comment]
the issue is that the `build context` should still be the root of the repo.. i.e. the equivalent of 

```
docker build -f docker/Dockerfile .
```

while what Docker Hub does is

```
cd docker
docker build .
```

**Comment by matthewfeickert** [comment]
@lukasheinrich Hm, I think I still don't follow what the problem is though with Docker Hub, as [this example repo](https://github.com/matthewfeickert/docker-build-alt-path-example) is able to [build correctly](https://hub.docker.com/r/matthewfeickert/docker-build-alt-path-example/builds/bupifasqd77cj9buy5dtpfw/) as long as the Dockerfile Location is set.

**Comment by matthewfeickert** [comment]
@lukasheinrich Oh, or are you saying that given that the Dockerfile that is currently used has this component to it

```Dockerfile
ADD . /code
RUN cd /code;pip install -e .
```

that this will fail given what Docker Hub does, but that Docker Cloud can somehow get around this?

**Comment by lukasheinrich** [comment]
@matthewfeickert yeah.. if you would add a file in the root of the repo and add a `ADD` statement it will not find it.

**Comment by matthewfeickert** [comment]
>  but I think the best practice is for the Dockerfile to use the code from the revision it is at (which might be a different branch etc)

Hm, I think I am misunderstanding something about what the difference is. With

```Dockerfile
ADD . /code
```

we're adding the contents of the `pyhf` GitHub repo into `/code/` in the image that is being built, but as we have the automated builds setup for pushes to only `master` this will only capture the state of the `master` branch (correct?). Isn't this identical to doing

```Dockerfile
RUN git clone --depth 1 https://github.com/diana-hep/pyhf.git code
```

?

Perhaps you're envisioning a different use case for the Docker image then I am? I was thinking that we would have the Docker image build only on `master` so that with ever push to `master` we also publish a working environment that someone can just pull down and use. Were you thinking of having it build and publish on every push to every branch so that it could serve as a drop in debugging environment?

I'm not doubting you on what is best practice ([for obvious reasons](https://github.com/reanahub/reana)), just curious about why having path dependent build is better.

**Comment by lukasheinrich** [comment]
both approaches work of course. Core issue is that we need *some* way to get our code into the image

1. via git
2. via a package manager
3. via some other network connection (e.g. `wget`-ing say the master tarball))
4. via the image build system provided by the container runtime  (i.e. `ADD`)

1-3 pose restrictions on the image in that it needs to have i.e. git, the pkg manager, wget or whatever, while 4 does not pose any restriction on the image.

if also has the added benefit that we can always locally test image building via 

```
docker build -t mytest -f docker/Dockerfile .
```
where this will lead evantually is to something like 

https://github.com/GoogleContainerTools/distroless

where the image is really reduced to the basics (no shell, no real operating system, no package manager, only the python runtime and dependent package)
